devito/interfaces.py

```diff
@@ -42,34 +130,57 @@ def initialize(self):
 
 
 class TimeData(DenseData):
-    # The code here is getting increasingly messy because python wants two types
-    # of constructors for everything. Since the parent class is Immutable, some
-    # constructor work needs to be moved into the __new__ method while some is in
-    # __init__. This makes it important to override both __new__ and __init__ in
-    # every child class.
-    def __init__(self, name, spc_shape, time_dim, time_order, save, dtype, pad_time=False):
-        if save:
-            time_dim = time_dim + time_order
+    """Data object for time-varying data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param shape: Shape of the spatial data grid
+    :param dtype: Data type of the buffered data
+    :param save: Save the intermediate results to the data buffer. Defaults
+                 to `False`, indicating the use of alternating buffers.
+    :param time_dim: Size of the time dimension that dictates the leading
+                     dimension of the data buffer if :param save: is True.
+    :param time_order: Order of the time discretization which affects the
+                       final size of the leading time dimension of the
+                       data buffer.
+
+    Note: The parameter :shape: should only define the spatial shape of the
+    grid. The temporal dimension will be inserted automatically as the
+    leading dimension, according to the :param time_dim:, :param time_order:
+    and whether we want to write intermediate timesteps in the buffer.
+    """
+
+    def __init__(self, *args, **kwargs):
+        if self._cached():
```

Comment:
There might be, but I haven't found it yet. However, I don't think we will create that many of them, and whenever a user/application developer wants to introduce new object types or sub-type ours, they should talk to the Devito dev team, or at least read the `SymbolicData` docstring.